framework/mesh/MeshGenerator/MeshGenerator_02_utils.cc

```diff
@@ -3,27 +3,54 @@
 namespace chi_mesh
 {
 
+// ###################################################################
+/**Broadcasts PIDs to other locations.*/
+void MeshGenerator::BroadcastPIDs(std::vector<int64_t>& cell_pids,
+                                  int root,
+                                  MPI_Comm communicator)
+{
+  size_t data_count = Chi::mpi.location_id == root ? cell_pids.size() : 0;
+
+  //======================================== Broadcast data_count to all
+  //                                         locations
+  MPI_Bcast(&data_count,   // buffer [IN/OUT]
+            1,             // count
+            MPI_UINT64_T,  // data type
+            root,          // root
+            communicator); // communicator
+
+  if (Chi::mpi.location_id != root) cell_pids.assign(data_count, 0);
+
+  //======================================== Broadcast partitioning to all
+  //                                         locations
+  MPI_Bcast(cell_pids.data(),             // buffer [IN/OUT]
+            static_cast<int>(data_count), // count
+            MPI_LONG_LONG_INT,            // data type
+            root,                         // root
+            communicator);                // communicator
+}
+
 // ###################################################################
 /**Determines if a cells needs to be included as a ghost or as a local cell.*/
-bool chi_mesh::MeshGenerator::CellHasLocalScope(
+bool MeshGenerator::CellHasLocalScope(
+  int location_id,
   const chi_mesh::UnpartitionedMesh::LightWeightCell& lwcell,
   uint64_t cell_global_id,
   const std::vector<std::set<uint64_t>>& vertex_subscriptions,
-  const std::vector<int64_t>& cell_partition_ids)
+  const std::vector<int64_t>& cell_partition_ids) const
 {
-  if (replicated_)
-    return true;
+  if (replicated_) return true;
   // First determine if the cell is a local cell
   int cell_pid = static_cast<int>(cell_partition_ids[cell_global_id]);
-  if (cell_pid == Chi::mpi.location_id) return true;
+  if (cell_pid == location_id) return true;
 
   // Now determine if the cell is a ghost cell
   for (uint64_t vid : lwcell.vertex_ids)
     for (uint64_t cid : vertex_subscriptions[vid])
     {
       if (cid == cell_global_id) continue;
       int adj_pid = static_cast<int>(cell_partition_ids[cid]);
```

Comment:
There is an interesting disparity in 3 places:
- We store `partition_id` on a cell as a `uint64_t`
- MPI uses `int` exclusively for ranks and rank count
- Finally, PETSc's partitioners work on `int64_t`

We store the partition_id as `uint64_t` because it simplified serialization and its one less bound to check since it unsigned, however, we could refactor to just `int` it would just require some small changes here and there.

Overall, the `int` on the mpi side is the determining factor. MPI just would start with an overflow of int so it is safe to downcast to it.